### PR TITLE
Add collector for SMB

### PR DESCRIPTION
**Pull request by aburtasov**
Added new collector for SMB Server
Collector collects 2 metrics:
- Current count of open files on the SMB Server(count of open files in shared folder)
- Current count of open session on the SMB Server(count of users connection to shared folder)

Thats metrics you can see on compmgmt.msc: Computer management -> Shared folders -> (Open files) and (Sessions)

### CODE REVIEW

**Comment by jkroepke** [comment]
@aburtasov Thanks for your PR. Do you have an action list for me to enable SMB Server on my system? I would like to test this.

File: pkg/collector/config.go [hunk 1]
```diff
@@ -107,6 +108,7 @@ type Config struct {
 	Time                           time.Config                            `yaml:"time"`
 	Vmware                         vmware.Config                          `yaml:"vmware"`
 	VmwareBlast                    vmware_blast.Config                    `yaml:"vmware_blast"`
+	Smb                            smb.Config                             `yaml:"smb"`
```
Comment (jkroepke) [inline pkg/collector/config.go:111]: please keep the list ordered.

File: pkg/collector/map.go [hunk 2]
```diff
@@ -111,6 +112,7 @@ var Map = map[string]types.CollectorBuilderWithFlags{
 	time.Name:                            time.NewWithFlags,
 	vmware.Name:                          vmware.NewWithFlags,
 	vmware_blast.Name:                    vmware_blast.NewWithFlags,
+	smb.Name:                             smb.NewWithFlags,
```
Comment (jkroepke) [inline pkg/collector/map.go:115]: please keep the list ordered.

File: pkg/collector/config.go [hunk 3]
```diff
@@ -163,4 +165,5 @@ var ConfigDefaults = Config{
 	Time:                           time.ConfigDefaults,
 	Vmware:                         vmware.ConfigDefaults,
 	VmwareBlast:                    vmware_blast.ConfigDefaults,
+	Smb:                            smb.ConfigDefaults,
```
Comment (jkroepke) [inline pkg/collector/config.go:168]: please keep the list ordered.

File: pkg/collector/collector.go [hunk 4]
```diff
@@ -136,6 +137,7 @@ func NewWithConfig(logger log.Logger, config Config) Collectors {
 	collectors[time.Name] = time.New(logger, &config.Time)
 	collectors[vmware.Name] = vmware.New(logger, &config.Vmware)
 	collectors[vmware_blast.Name] = vmware_blast.New(logger, &config.VmwareBlast)
+	collectors[smb.Name] = smb.New(logger, &config.Smb)
```
Comment (jkroepke) [inline pkg/collector/collector.go:140]: Please call the collector ordered.

**Comment by jkroepke** [comment]
@aburtasov I could see a older PR which also introduces a lot of more SMB Share metrics. Do you feel it makes sense to include them?

https://github.com/prometheus-community/windows_exporter/pull/501/files

**Comment by aburtasov** [comment]
> @aburtasov Thanks for your PR. Do you have an action list for me to enable SMB Server on my system? I would like to test this.

@jkroepke Hey! Here a good article how to enable SMB: https://www.themakersphere.com/setup-smb-file-share-on-windows/

File: pkg/collector/config.go (hunk 1 above)
Comment (aburtasov) [reply pkg/collector/config.go]: fixed that in new commit

File: pkg/collector/map.go (hunk 2 above)
Comment (aburtasov) [reply pkg/collector/map.go]: fixed that in new commit

File: pkg/collector/config.go (hunk 3 above)
Comment (aburtasov) [reply pkg/collector/config.go]: fixed that in new commit

File: pkg/collector/collector.go (hunk 4 above)
Comment (aburtasov) [reply pkg/collector/collector.go]: fixed that in new commit

**Comment by aburtasov** [comment]
> @aburtasov I could see a older PR which also introduces a lot of more SMB Share metrics. Do you feel it makes sense to include them?
> 
> https://github.com/prometheus-community/windows_exporter/pull/501/files

Hard to say. Our Windows Server enginners checks only this 2 metrics that i added to smb collector

**Comment by aburtasov** [comment]
@jkroepke I added more description about smb collector in PR-comment.

**Review by jkroepke — APPROVED** [review]
LGTM

File: pkg/collector/smb/smb.go [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+		for _, cname := range smbAllCollectorNames {
+			fmt.Printf("%-32s %-32s\n", cname, collectorDesc[cname])
+		}
+		os.Exit(0)
+	}
+
+	if *c.smbCollectorsEnabled == "" {
+		for _, collectorName := range smbAllCollectorNames {
+			c.enabledCollectors = append(c.enabledCollectors, collectorName)
+		}
```
[the remark this answers is not in the document]
Comment (aburtasov) [reply pkg/collector/smb/smb.go:123]: Hmm, in you case there will be an empty slice, as i see

**Comment by jkroepke** [comment]
Works, thanks!

**Pull request merged**